src/connectivity/mod.rs

```diff
@@ -266,6 +270,63 @@ pub fn is_weakly_connected(graph: &digraph::PyDiGraph) -> PyResult<bool> {
     Ok(weakly_connected_components(graph)[0].len() == graph.graph.node_count())
 }
 
+/// Check if the graph is semi connected
+///
+/// :param PyDiGraph graph: The graph to check if it is semi connected
+///
+/// :returns: Whether the graph is semi connected or not
+/// :rtype: bool
+///
+/// :raises NullGraph: If an empty graph is passed in
+#[pyfunction]
+#[pyo3(text_signature = "(graph, /)")]
+pub fn is_semi_connected(graph: &digraph::PyDiGraph) -> PyResult<bool> {
+    if graph.graph.node_count() == 0 {
+        return Err(NullGraph::new_err("Invalid operation on a NullGraph"));
+    }
+
+    let mut temp_graph = DiGraph::new();
+    let mut node_map = Vec::new();
+
+    for _node in graph.graph.node_indices() {
+        node_map.push(temp_graph.add_node(()));
+    }
+
+    for edge in graph.graph.edge_indices() {
+        let (source, target) = graph.graph.edge_endpoints(edge).unwrap();
+        temp_graph.add_edge(node_map[source.index()], node_map[target.index()], ());
+    }
+
+    let condensed = condensation(temp_graph, false);
+    let condensed_digraph = DiGraph::from(condensed);
+
+    let found_path: RwLock<bool> = RwLock::new(true);
+    
+    let topo_sort = match toposort(&condensed_digraph, None){
+        Ok(topo_sort) => topo_sort,
+        Err(_) => {
+            let mut dfs = DfsPostOrder::new(&condensed_digraph, NodeIndex::new(0));
+            let mut nodes = Vec::new();
+            while let Some(node) = dfs.next(&condensed_digraph){
+                nodes.push(node);
+            }
+            nodes.reverse();
+            nodes
+        },
+    };
+
+    let pairs: Vec<_> = topo_sort.windows(2).map(|window| (window[0], window[1])).collect();
+    pairs.into_par_iter().for_each(|(u,v)| {
+        if !(petgraph::algo::has_path_connecting(&condensed_digraph, u, v, None)){
```

Comment:
Thinking more deeply about this we don't need `has_path_connecting`, I think just checking if the edge exists is good enough